### PR TITLE
Rewritten Count-Min Sketch

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

File: src/rm_cms.c [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    char *cRedisStr = (char *)RedisModule_StringPtrLen(redisStr, &len);
+    return strncasecmp(cRedisStr, str, len) == 0;
+}
+
+static int GetCMS(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+                    CMSketch **cms, int mode) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_ReplyWithError(ctx, "CMS: key does not exist");
+        return REDISMODULE_ERR;
```
Comment (itamarhaber) [inline src/rm_cms.c]: This should be REDISMODULE_OK. _ERR is reserved for real errors and, afair, isn't really used atm. Same goes for the next return.

File: src/rm_cms.c [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    return REDISMODULE_OK;
+}
+
+static int CreateCmsKey(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+        long long width, long long depth, CMSketch **cms, RedisModuleKey **key) {
+    if(*key == NULL) {
+        *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+    }
+
+    RedisModule_RetainString(ctx, keyName);
```
Comment (itamarhaber) [inline src/rm_cms.c]: Why is that needed? And if it is, where is the free?

File: src/cms.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#endif
+
+typedef __uint128_t long128;
+
+typedef struct
+{
+    size_t width;
+    size_t depth;
+    size_t *array;    
+    long128 counter; // might be used for top k results
```
Comment (itamarhaber) [inline src/cms.h]: Not sure what this is for, but better leave out until it is needed.

File: src/rm_cms.c [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    if(argc != 4)
+        return RedisModule_WrongArity(ctx);
+
+    CMSketch *cms;
+    long long width, depth;
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+    if(parseCreateArgs(ctx, argv, argc, &width, &depth) != REDISMODULE_OK)
+        return REDISMODULE_ERR;
```
Comment (itamarhaber) [inline src/rm_cms.c]: Need to close the key

File: src/rm_cms.c [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+    if(RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx,"CMS: key already exists");
+    }
+
+    CreateCmsKey(ctx, keyName, width, depth, &cms, &key);
+    RedisModule_CloseKey(key);
+
+    RedisModule_Log(ctx, "info", "created new count-min sketch");
```
Comment (itamarhaber) [inline src/rm_cms.c]: This is being "big head" - I like that :) If that is the case, why not include more details like params and mebbe key name? That said, I'm not sure that the log message is really needed as we don't usually log key lifetime events.

File: src/rm_cms.c [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ);
+    CMSketch *cms;
+    CMSPair *pairArray;
+    parseIncrByArgs(ctx, argv, argc, &pairArray, pairCount);
+
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        CreateCmsKey(ctx, keyName, DEFAULT_WIDTH, DEFAULT_DEPTH, &cms, &key);
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
```
Comment (itamarhaber) [inline src/rm_cms.c]: Missing RM_CloseKey

File: src/rm_cms.c [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    size_t len;
+    char *cRedisStr = (char *)RedisModule_StringPtrLen(redisStr, &len);
+    return strncasecmp(cRedisStr, str, len) == 0;
+}
+
+static int GetCMS(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+                    CMSketch **cms, int mode) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_ReplyWithError(ctx, "CMS: key does not exist");
```
Comment (itamarhaber) [inline src/rm_cms.c]: Missing RM_CloseKey

File: src/rm_cms.c [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+        return REDISMODULE_ERR;
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+        return REDISMODULE_ERR;
```
Comment (itamarhaber) [inline src/rm_cms.c]: Missing RM_CloseKey

File: src/rm_cms.c [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+int CMSketch_query(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+
+    if(argc < 3) {
+        return RedisModule_WrongArity(ctx);
+    }
+
+    CMSketch *cms = NULL;
+    if(GetCMS(ctx, argv[1], &cms, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_ERR;
```
Comment (itamarhaber) [inline src/rm_cms.c]: Missing RM_CloseKey, no need to return REDISMODULE_ERR

File: src/rm_cms.c [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+void CMSFree(void *value) { 
+    CMS_Destroy(value);
+}
+
+size_t CMSMemUsage(const void *value) {
+    CMSketch *cms = (CMSketch *)value;
+    return sizeof(cms) + cms->width * cms->depth * sizeof(size_t);
+}
+
+void AofTemp(RedisModuleIO *aof, RedisModuleString *key, void *value) {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/rm_cms.c]: Remains

File: src/rm_cms.c [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+        return REDISMODULE_ERR;
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+        return REDISMODULE_ERR;
+    } 
+    *cms = RedisModule_ModuleTypeGetValue(key);
+    return REDISMODULE_OK;
+}
+
+static int CreateCmsKey(RedisModuleCtx *ctx, RedisModuleString *keyName,
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/rm_cms.c]: Changed

File: src/cms.h [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#ifndef RM_CMS_H
+#define RM_CMS_H
+
+#include <stddef.h>
+
+#include "../contrib/xxhash.h"
+
+// #define REDIS_MODULE_TARGET
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/cms.h]: Done

File: src/test.c [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+int visualTest();
+
+int main()
+{
+    visualTest();
+
+    return 0;
+}
+
+int visualTest() {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/test.c]: WIP

File: src/cms.c [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+#include <assert.h>         // assert
+
+#include "cms.h"
+
+CMSketch *NewCMSketch(size_t width, size_t depth) {
+    assert(width > 0);
+    assert(depth > 0);
+
+    CMSketch *cms = NULL;
+    if((cms = (CMSketch *)CMS_CALLOC(1, sizeof(CMSketch))) == NULL) {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/cms.c]: I see some mixed code about this. Since it isn't called much, it might be OK.
This is a good opportunity to get out of the bubble of Infinity to the real world :)

File: src/rm_cms.c (hunk 1 above)
Comment (ashtul) [reply src/rm_cms.c]: This is for internal use. The function which calls it returns _OK

File: src/cms.h (hunk 3 above)
Comment (ashtul) [reply src/cms.h]: I thought to make space for a larger global counter but ended up not using.
Cleaned!

File: src/rm_cms.c (hunk 9 above)
Comment (ashtul) [reply src/rm_cms.c]: The closekey isn't needed as I called RedisModule_AutoMemory

File: src/rm_cms.c (hunk 5 above)
Comment (ashtul) [reply src/rm_cms.c]: I need to learn more about it.
it was copied from reference w/o enough understanding of why and when.

File: src/rm_cms.c (hunk 2 above)
Comment (ashtul) [reply src/rm_cms.c]: Not sure :(
It was referenced as black magic. 
Will study further.

File: src/rm_cms.c (hunk 1 above)
Comment (itamarhaber) [reply src/rm_cms.c]: If for internal use just return 0 or 1, or make your own defines... imo. The current is kind of confusing (to me at least).

File: src/rm_cms.c (hunk 2 above)
Comment (itamarhaber) [reply src/rm_cms.c]: Always keep learning.

File: src/rm_cms.c (hunk 9 above)
Comment (itamarhaber) [reply src/rm_cms.c]: 👍

File: src/rm_cms.c (hunk 1 above)
Comment (ashtul) [reply src/rm_cms.c]: 👍

File: Makefile [hunk 15]
```diff
@@ -28,8 +28,12 @@ MODULE_OBJ = $(SRCDIR)/rebloom.o
 MODULE_SO = $(ROOT)/rebloom.so
 
 DEPS = $(ROOT)/contrib/MurmurHash2.o \
+	   $(ROOT)/contrib/xxhash.o \
```
Comment (mnunberg) [inline Makefile]: If possible, please use an existing hash (murmur or whatever..)

File: contrib/xxhash.h [hunk 16]
```diff
@@ -0,0 +1,156 @@
+/*
```
[the remark this answers is not in the document]
Comment (mnunberg) [reply contrib/xxhash.h]: submodules are a pain to deal with, unless we really use the entire library

File: src/cms.c [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+void CMS_Destroy(CMSketch *cms) {
+    assert(cms);
+
+    CMS_FREE(cms->array);
+    cms->array = NULL;
+
+    CMS_FREE(cms);
+}
+
+void CMS_IncrBy(CMSketch *cms, const char *item, size_t value) {
```
Comment (mnunberg) [inline src/cms.c]: Accept a buffer and a length, rather than assume a NUL-terminated string. e.g. `const char *item, size_t itemlen`, and then pass `itemlen` to the hash function.

File: src/cms.c [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    assert(item);
+
+    for(size_t i = 0; i < cms->depth; ++i) {
+        size_t hash = XXH64(item, strlen(item), i);
+        cms->array[(hash % cms->width) + (i * cms->width)] += value;
+    }
+    cms->counter += value;
+}
+
+size_t CMS_Query(CMSketch *cms, const char *item) {
```
Comment (mnunberg) [inline src/cms.c]: and length

File: src/cms.c [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    assert(cms);
+    assert(item >= 0);
+
+    size_t temp = 0, res = ~0;
```
Comment (mnunberg) [inline src/cms.c]: `~0LLU`, `-1`, `SIZE_MAX`, or somesuch. I'm not 100% sure that `~0` does what you think it does. I've had odd moments where it didn't set all the bits (:o)

File: src/cms.c [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+
+    for(size_t i = 0; i < cms->depth; ++i) {
```
Comment (mnunberg) [inline src/cms.c]: `for (size_t` -- space between for and opening `(`

File: src/cms.c [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        size_t hash = XXH64(item, strlen(item), i);
+        temp = cms->array[(hash % cms->width) + (i * cms->width)];
+        if(temp < res) res = temp;
```
Comment (mnunberg) [inline src/cms.c]: space, braces in if block

File: src/cms.c [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    return res;
+}
+
+void CMS_Merge(CMSketch *dest, size_t quantity, CMSketch **src, long long *weight) {
```
Comment (mnunberg) [inline src/cms.c]: `src` is `const` and should be declared as such. `weight` is also const

File: src/cms.h [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+#ifdef REDIS_MODULE_TARGET // should be in .h or .c
+#define CMS_CALLOC(count, size) RedisModule_Calloc(count, size)
+#define CMS_FREE(ptr) RedisModule_Free(ptr)
+#else
+#define CMS_CALLOC(count, size) calloc(count, size)
+#define CMS_FREE(ptr) free(ptr)
+#endif
+
+typedef struct
+{
```
Comment (mnunberg) [inline src/cms.h]: brace on same line as typedef

File: src/rm_cms.c [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+#include "cms.h"
+
+RedisModuleType *CMSketchType;
+
+typedef struct {
+    const char *key;
+    long long value;
+} CMSPair;
+
+static int CompareCStringToRedisString(char *str, RedisModuleString *redisStr) {
```
Comment (mnunberg) [inline src/rm_cms.c]: `const char*`

Comment (mnunberg) [inline src/rm_cms.c]: Don't use CamelCase for functions that aren't objectish, use snakeCase. Also, I believe we have this exact function in rmutil, so you can borrow from there.

Also, I like shorter names. It gets difficult to read in code if you have giant names followed by other giant names.

e.g. `rstringCmp`

File: src/rm_cms.c (hunk 1 above)
Comment (mnunberg) [reply src/rm_cms.c]: In RediSearch and elsewhere, we use REDISEARCH_OK/REDISEARCH_ERR as internal return codes all the time. Nothing wrong with that as long as it's consistent and documented. I believe they expand to the same 0 and 1 anyway, though.

File: src/rm_cms.c (hunk 2 above)
Comment (mnunberg) [reply src/rm_cms.c]: Retain/Free work as reference counter increments/decrements

File: src/rm_cms.c [hunk 25]
```diff
@@ -0,0 +1,328 @@
...
+        return REDISMODULE_ERR;
+    }
+
+    return REDISMODULE_OK;
+}
+
+static int parseCreateArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
+                    long long *width, long long *depth) {
+    if(CompareCStringToRedisString("cms.initbydim", argv[0]) == 1) {
+        if((RedisModule_StringToLongLong(argv[2], width) != REDISMODULE_OK)
```
Comment (mnunberg) [inline src/rm_cms.c]: Wrap on same line. We use 100 column line widths

File: src/rm_cms.c [hunk 26]
```diff
@@ -0,0 +1,328 @@
...
+        }
+        *width = ceil(2 / err);
+        *depth = ceil(log10f(prob) / log10f(0.5));
+    }
+    
+    return REDISMODULE_OK;
+}
+
+int CMSketch_create(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    if(argc != 4)
```
Comment (mnunberg) [inline src/rm_cms.c]: brace

File: src/rm_cms.c [hunk 27]
```diff
@@ -0,0 +1,328 @@
...
+            return REDISMODULE_ERR;
+        }
+        *width = ceil(2 / err);
+        *depth = ceil(log10f(prob) / log10f(0.5));
+    }
+    
+    return REDISMODULE_OK;
+}
+
+int CMSketch_create(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
```
Comment (mnunberg) [inline src/rm_cms.c]: why `create` and not `Create`?

File: src/rm_cms.c [hunk 28]
```diff
@@ -0,0 +1,328 @@
...
+    RedisModule_ReplyWithArray(ctx, itemCount);
+    for(int i = 0; i < itemCount; ++i) {
+        const char *str = RedisModule_StringPtrLen(argv[2 + i], &length);
+        RedisModule_ReplyWithLongLong(ctx, CMS_Query(cms, str));
+    }
+
+    return REDISMODULE_OK;
+}
+
+static int parseMergeArgs(RedisModuleCtx *ctx, RedisModuleString **argv,
```
Comment (mnunberg) [inline src/rm_cms.c]: suggest that you use a structur e.g. `typedef struct { ... } mergeParams`

File: src/rm_cms.c [hunk 29]
```diff
@@ -0,0 +1,328 @@
...
+    RedisModuleTypeMethods tm = {
+            .version = REDISMODULE_TYPE_METHOD_VERSION,
+            .rdb_load = CMSRdbLoad,
+            .rdb_save = CMSRdbSave,
+            .aof_rewrite = RMUtil_DefaultAofRewrite, 
+            .mem_usage = CMSMemUsage,
+            .free = CMSFree
+        };
+
+    CMSketchType = RedisModule_CreateDataType(ctx, "CMSk-TYPE", CMS_ENC_VER, &tm);
```
Comment (mnunberg) [inline src/rm_cms.c]: I believe there are some naming restrictions on the module itself, I'd just use `cms`

File: tests/cms.py [hunk 30]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python
+from rmtest import ModuleTestCase
+from redis import ResponseError
+import sys
+
+if sys.version >= '3':
+    xrange = range
+
+class RebloomTestCase(ModuleTestCase('../rebloom.so')):
```
Comment (mnunberg) [inline tests/cms.py]: Please rename this

File: tests/cms.py [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+        self.cmd('cms.initbydim', 'large_5', '2000', '10')
+        self.cmd('cms.initbydim', 'large_6', '2000', '10')
+
+        self.assertOk(self.cmd('cms.merge', 'small_3', 2, 'small_1', 'small_2'))
+        self.assertOk(self.cmd('cms.merge', 'large_6', 2, 'large_4', 'large_5'))
+        self.assertRaises(ResponseError, self.cmd, 'cms.merge', 'small_3', 2,
+                                    'small_2', 'large_5')
+
+'''
+    def test_dump_and_load(self):
```
Comment (mnunberg) [inline tests/cms.py]: remove

File: src/rm_cms.c (hunk 2 above)
Comment (ashtul) [reply src/rm_cms.c]: Removed

File: src/rm_cms.c (hunk 5 above)
Comment (ashtul) [reply src/rm_cms.c]: Removed for now

File: src/rm_cms.c (hunk 29 above)
Comment (ashtul) [reply src/rm_cms.c]: It must be 9 chars

File: Makefile (hunk 15 above)
Comment (ashtul) [reply Makefile]: From my understanding, it is used in the team.
Benchmarks show it is considerably faster.

File: src/cms.c (hunk 22 above)
Comment (ashtul) [reply src/cms.c]: weight is always const.
src might also be dest so it a matter of what would be more clear for developers.

File: src/rm_cms.c (hunk 1 above)
Comment (ashtul) [reply src/rm_cms.c]: Added INNER_ERROR with both RedisModule_ReplyWithError and ERROR

File: src/rm_cms.h [hunk 32]
```diff
@@ -0,0 +1,13 @@
+#ifndef CMS_MODULE_H
+#define CMS_MODULE_H
+
+#include <redismodule.h>
+
+#define DEFAULT_WIDTH 2000
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/rm_cms.h]: Added todo comment

File: tests/cms.py (hunk 30 above)
Comment (ashtul) [reply tests/cms.py]: Good catch

File: Makefile (hunk 15 above)
Comment (ashtul) [reply Makefile]: After running my own benchmark, MurmurHash came on top some I remove XXHash.

**Comment by ashtul** [comment]
Some major change including to the API.
Instead of err and prob, we require 'n' as well (I guess I should change that to 'size').
Functions to calculate width and depth out of the above have changed.

trying to figure out the circleci error.

File: Makefile (hunk 15 above)
Comment (mnunberg) [reply Makefile]: xxhash is still here.. but I did some research and ppl claim it's faster, so... why not.

File: src/cms.c [hunk 33]
```diff
@@ -17,11 +17,23 @@ CMSketch *NewCMSketch(size_t width, size_t depth) {
     cms->width = width;
     cms->depth = depth;
     cms->counter = 0;
-    cms->array = CMS_CALLOC(width * depth, sizeof(size_t));
+    cms->array = CMS_CALLOC(width * depth, sizeof(uint));
```
Comment (mnunberg) [inline src/cms.c]: `uint` is not a standardized type. Please use either `unsigned` or `uint32_t`

File: src/cms.c [hunk 34]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
     CMS_FREE(cms);
 }
 
-void CMS_IncrBy(CMSketch *cms, const char *item, size_t value) {
+void CMS_IncrBy(CMSketch *cms, const char *item, size_t strlen, size_t value) {
```
Comment (mnunberg) [inline src/cms.c]: `itemlen`; it's confusing to see `strlen` as a variable name.

File: src/cms.c [hunk 35]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
     assert(cms);
     assert(item);
 
     for(size_t i = 0; i < cms->depth; ++i) {
-        size_t hash = XXH64(item, strlen(item), i);
+        uint32_t hash = MurmurHash2(item, strlen, i);
```
Comment (mnunberg) [inline src/cms.c]: Perhaps it's better to make this into a macro: `CMS_HASH(itm, len, iter)`

File: src/cms.c [hunk 36]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
     cms->counter += value;
 }
 
-size_t CMS_Query(CMSketch *cms, const char *item) {
+size_t CMS_Query(CMSketch *cms, const char *item, size_t strlen) {
     assert(cms);
     assert(item >= 0);
 
-    size_t temp = 0, res = ~0;
+    size_t temp = 0, res = -1;
```
Comment (mnunberg) [inline src/cms.c]: I usually do `(size_t)-1`.. I'm not sure that's needed, but it makes me feel better :)

File: src/cms.c [hunk 37]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
 
     for(size_t i = 0; i < cms->depth; ++i) {
-        size_t hash = XXH64(item, strlen(item), i);
+        uint hash = MurmurHash2(item, strlen, i);
```
Comment (mnunberg) [inline src/cms.c]: `uint32_t`

File: src/cms.c [hunk 38]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
         temp = cms->array[(hash % cms->width) + (i * cms->width)];
-        if(temp < res) res = temp;
+        if(temp < res) {
```
Comment (mnunberg) [inline src/cms.c]: Formatting. I suggest you use `clang-format` (already an option in vscode)

File: src/cms.c [hunk 39]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
+        cmsCount += src[i]->counter * weights[i];
     }
-    dest->counter = tempTotal;
+    dest->counter = cmsCount;
 }
  
 void CMS_Print(const CMSketch *cms) {
     assert(cms);
 
     for(int i = 0; i < cms->depth; ++i) {
```
Comment (mnunberg) [inline src/cms.c]: formatting..

File: src/cms.c [hunk 40]
```diff
@@ -31,66 +43,84 @@ void CMS_Destroy(CMSketch *cms) {
...
-            printf("%lu\t", cms->array[(i * cms->width) + j]);
+            printf("%d\t", cms->array[(i * cms->width) + j]);
         }
         printf("\n");
     }
-    printf("\tCounter is %lu\n", (size_t)cms->counter);
+    printf("\tCounter is %lu\n", cms->counter);
+}
+
+size_t CMS_cardinality(CMSketch *cms) {
```
Comment (mnunberg) [inline src/cms.c]: CMS_GetCardinality

File: src/cms.h [hunk 41]
```diff
@@ -13,19 +13,45 @@
 #define CMS_FREE(ptr) free(ptr)
 #endif
 
-typedef struct
-{
+typedef unsigned int uint;
```
Comment (mnunberg) [inline src/cms.h]: :(

Comment (mnunberg) [reply src/cms.h]: Maybe something like `typedef uint32_t CMSHashType` ?

File: src/rm_cms.c [hunk 42]
```diff
@@ -25,14 +30,11 @@ static int GetCMSKey(RedisModuleCtx *ctx, RedisModuleString *keyName,
                     CMSketch **cms, int mode) {
     RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
     if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
-        RedisModule_ReplyWithError(ctx, "CMS: key does not exist");
-        return REDISMODULE_ERR;
+        INNER_ERROR("CMS: key does not exist");
```
Comment (mnunberg) [inline src/rm_cms.c]: This will leak `key`. If you see what we do in a lot of other modules, at the end of the function we have a `done` block where we apply `goto`.

File: src/rm_cms.c [hunk 43]
```diff
@@ -54,72 +55,123 @@ static int CreateCMSKey(RedisModuleCtx *ctx, RedisModuleString *keyName,
 
 static int parseCreateArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
                     long long *width, long long *depth) {
-    if(CompareCStringToRedisString("cms.initbydim", argv[0]) == 1) {
+    if(argc == 4) {
```
Comment (mnunberg) [inline src/rm_cms.c]: You can use `args.{c,h}` from rmutil, might make it easier to parse

File: src/rm_cms.c [hunk 44]
```diff
@@ -54,72 +55,123 @@ static int CreateCMSKey(RedisModuleCtx *ctx, RedisModuleString *keyName,
...
-            return REDISMODULE_ERR;
+        long long n = 0;
+        double overEst = 0, prob = 0;
+        if((RedisModule_StringToLongLong(argv[2], &n) != REDISMODULE_OK) ||
+                                (n < 1)) {
+            INNER_ERROR("CMS: invalid n value");
+        } 
+        if((RedisModule_StringToDouble(argv[3], &overEst) != REDISMODULE_OK) ||
+                                overEst <= 0 || overEst >= 1) {
+            INNER_ERROR("CMS: invalid over estimation value");
```
Comment (mnunberg) [inline src/rm_cms.c]: overestimation (one word)

File: src/rm_cms.c [hunk 45]
```diff
@@ -203,33 +249,34 @@ static int parseMergeArgs(RedisModuleCtx *ctx, RedisModuleString **argv,
     size_t depth = (*dest)->depth;
 
     for(int i = 0; i < numKeys; ++i) {
-        if(pos > 0) {
-            if(RedisModule_StringToLongLong(argv[3 + numKeys + 1 + i], &weights[i]) !=
+        if(pos == -1) { weights[i] = 1; }
```
Comment (mnunberg) [inline src/rm_cms.c]: need newlines

File: src/rm_cms.c [hunk 46]
```diff
@@ -203,33 +249,34 @@ static int parseMergeArgs(RedisModuleCtx *ctx, RedisModuleString **argv,
...
                                                 REDISMODULE_OK) {
-                RedisModule_ReplyWithError(ctx, "CMS: invalid weight value");
-                return REDISMODULE_ERR;
-            }
-        } else { weights[i] = 1; }
+                INNER_ERROR("CMS: invalid weight value");
+        }
         if(GetCMSKey(ctx, argv[3 + i], &cmsArray[i], REDISMODULE_READ) != REDISMODULE_OK) {
-           return REDISMODULE_ERR;
+           return ERROR;
```
Comment (mnunberg) [inline src/rm_cms.c]: Just use `REDISMODULE_ERR`

File: src/cms.c (hunk 36 above)
Comment (ashtul) [reply src/cms.c]: I like this

File: src/cms.c (hunk 40 above)
Comment (ashtul) [reply src/cms.c]: It isn't yet integrated in.
For now, I'll probably add this to cms.info

File: src/cms.h (hunk 41 above)
Comment (ashtul) [reply src/cms.h]: It isn't just for the hash but the array itself.
I changed it to uint32_t as requested

File: src/rm_cms.c (hunk 42 above)
Comment (ashtul) [reply src/rm_cms.c]: I use RedisModule_AutoMemory so this isn't an issue.

I was very surprised to see goto but then read the one case to use it is cleanup

File: src/cms.c (hunk 39 above)
Comment (ashtul) [reply src/cms.c]: ?

File: src/cms.c (hunk 38 above)
Comment (ashtul) [reply src/cms.c]: Nice one.

File: src/rm_cms.c (hunk 43 above)
Comment (ashtul) [reply src/rm_cms.c]: used RMUtil_ArgIndex

File: src/.clang-format [hunk 47]
```diff
@@ -0,0 +1,12 @@
...
+PointerAlignment: Right
+TabWidth: 4
+IndentWidth: 4
+AccessModifierOffset: 0
+ColumnLimit: 120
+NamespaceIndentation: All
+AlignTrailingComments: true
+AllowAllParametersOfDeclarationOnNextLine: true
+AlwaysBreakTemplateDeclarations: true
+UseTab: Always
```
Comment (mnunberg) [inline src/.clang-format]:
```
IndentWidth: 4
ColumnLimit: 100
SortIncludes: false
AlignEscapedNewlinesLeft: false
SpacesBeforeTrailingComments: 1
```

Is the contents of the current file. I thought we had it in the repo, but it appears we don't.

We're not using tabs in this project

File: src/rm_cms.c (hunk 42 above)
Comment (mnunberg) [reply src/rm_cms.c]: `goto` isn't evil; it's specifically meant (and used) for cases like these. In any event, at least specify this in a comment.

I personally am not a fan of AutoMemory, especially in more complex commands; but I suppose cms is fairly straight forward

File: src/rm_cms.c (hunk 28 above)
Comment (ashtul) [reply src/rm_cms.c]: Done

Comment (ashtul) [reply src/rm_cms.c]: Done

**Comment by ashtul** [comment]
@mnunberg @itamarhaber @gkorland 
Changes requested by Mark implemented.
I believe we should drop default size (used to be 2000) altogether or at least make it much larger.

File: docs/CountMinSketch_Commands.md [hunk 48]
```diff
@@ -0,0 +1,50 @@
...
+CMS.INITBYDIM key width depth
+```
+
+Initializes a CMS to the dimensions specified by width and depth.
+
+## Parameters:
+
+* **key**: Key under which the sketch is to be found.
+* **width**: Number of counters kept in each array.
+* **Depth**: Number of arrays.
```
Comment (itamarhaber) [inline docs/CountMinSketch_Commands.md]: should be 'depth' (lowercase D)

File: src/.clang-format [hunk 49]
```diff
@@ -0,0 +1,6 @@
+
+IndentWidth: 4
+ColumnLimit: 100
```
Comment (itamarhaber) [inline src/.clang-format]: In Redis core it is 80, but who cares, right? ;P

File: src/rm_cms.c (hunk 29 above)
Comment (itamarhaber) [reply src/rm_cms.c]: IIRC @ashtul is correct
